File: 2024-04-15/Same_Tree.java
//  100. Same Tree
//
//  https://leetcode.com/problems/same-tree/description/

public class Same_Tree {
    public static boolean isSameTree(TreeNode p, TreeNode q) {
        if(p == null && q == null) return true;
        if(p == null || q == null) return false;
        if(p.val != q.val) return false;
        if(p.left == null && p.right == null && q.left == null && q.right == null) return true;

        return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
    }
    public static void main(String[] args) {
        Integer[] arr1 = new Integer[]{1,null,2};
        Integer[] arr2 = new Integer[]{1,null,2};

        Tree tree1 = new Tree();
        TreeNode root1 = tree1.arrayToTree(arr1);

        Tree tree2 = new Tree();
        TreeNode root2 = tree2.arrayToTree(arr2);

        System.out.println(isSameTree(root1, root2));
    }
}

File: 2026-02-13/Same_Tree.py
from typing import Optional, List
from Tree import TreeNode, Tree


class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        elif not p or not q or p.val != q.val:
            return False

        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)


sol = Solution()
tree = Tree()
node = tree.array_to_tree([1,2,3])
node2 = tree.array_to_tree([1,2,3])
print(sol.isSameTree(node, node2))
